Document new Roslyn Analyzer Rules for Wilson Token Validation (#25911)
dotnet/docs: docs/fundamentals/code-analysis/quality-rules/ca5405.md

---
title: "CA5045: Do not always skip token validation in delegates (code analysis)"
description: Provides information about code analysis rule CA5405, including causes, how to fix violations, and when to suppress it.
ms.date: 09/01/2021
ms.topic: reference
ms.author: timhann
f1_keywords:
  - "CA5405"
---

> # CA5359: Do not always skip token validation in delegates

Reviewer: ```suggestion
# CA5405: Do not always skip token validation in delegates
```

| | Value |
|-|-|
| **Rule ID** |CA5405|
| **Category** |[Security](security-warnings.md)|
| **Fix is breaking or non-breaking** |Non-breaking|

## Cause

The callback assigned to <xref:Microsoft.IdentityModel.Tokens.AudienceValidator>  or <xref:Microsoft.IdentityModel.Tokens.LifetimeValidator> always returns `true`.

## Rule description

> By setting critical TokenValidationParameter validations delegates to always return true, important authentication safeguards are disabled which can lead to tokens from any issuer or expired tokens being wrongly validated.

Reviewer: ```suggestion
By setting critical `TokenValidationParameter` validation delegates to always return `true`, important authentication safeguards are disabled. Disabling safeguards can lead to incorrect validation of tokens from any issuer or expired tokens.
```

> More details about best practices for token validation can be found on the [library's wiki](https://aka.ms/wilson/tokenvalidation).

Reviewer: ```suggestion
For more information about best practices for token validation, see the [library's wiki](https://aka.ms/wilson/tokenvalidation).
```

## How to fix violations

> - Improve the logic of the delegate so not all code paths return `true` which effectively disables that type of validation.

Reviewer: ```suggestion
- Improve the logic of the delegate so not all code paths return `true`, which effectively disables that type of validation.
```

> - `SecurityTokenInvalidAudienceException` or `SecurityTokenInvalidLifetimeException` (respectively) can be thrown in failure cases for when you want to fail validation and have other cases pass by returning `true`.

Reviewer: ```suggestion
- Throw `SecurityTokenInvalidAudienceException` or `SecurityTokenInvalidLifetimeException` in failure cases when you want to fail validation and have other cases pass by returning `true`.
```

## When to suppress warnings

> In some very specific cases where you're utilizing the delegate for additional logging and it's for token types where, for special reasons, the specifc type of validation is not needed it may make sense to supress this warning. Before you disable this validation, be sure you have fully thought through the security implications. The [library's wiki](https://aka.ms/wilson/tokenvalidation) can prove useful when thinking about the tradeoffs.

Reviewer: ```suggestion
In some specific cases where you're utilizing the delegate for additional logging and it's for token types where the specific type of validation is not needed, it may make sense to suppress this warning. Before you disable this validation, be sure you have fully thought through the security implications. For information about the trade-offs, see the [token validation library's wiki](https://aka.ms/wilson/tokenvalidation).
```
